x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -565,133 +475,177 @@ export class Authenticator {
+   * belongs to the provider that is not available.
+   * @param request Request instance.
    */
-  private async getSessionValue(sessionStorage: SessionStorage<ProviderSession>) {
-    const sessionValue = await sessionStorage.get();
+  private async getSessionValue(request: KibanaRequest) {
+    const existingSessionValue = await this.session.get(request);
 
-    // If we detect that session is in incompatible format or for some reason we have a session
-    // stored for the provider that is not available anymore (e.g. when user was logged in with one
-    // provider, but then configuration has changed and that provider is no longer available), then
-    // we should clear session entirely.
+    // If we detect that for some reason we have a session stored for the provider that is not
+    // available anymore (e.g. when user was logged in with one provider, but then configuration has
+    // changed and that provider is no longer available), then we should clear session entirely.
     if (
-      sessionValue &&
-      (isLegacyProviderSession(sessionValue) ||
-        this.providers.get(sessionValue.provider.name)?.type !== sessionValue.provider.type)
+      existingSessionValue &&
+      this.providers.get(existingSessionValue.provider.name)?.type !==
+        existingSessionValue.provider.type
     ) {
-      sessionStorage.clear();
+      this.logger.warn(
+        `Attempted to retrieve session for the "${existingSessionValue.provider.type}/${existingSessionValue.provider.name}" provider, but it is not configured.`
+      );
+      await this.session.clear(request);
       return null;
     }
 
-    return sessionValue;
+    return existingSessionValue;
   }
 
-  private updateSessionValue(
-    sessionStorage: SessionStorage<ProviderSession>,
+  /**
+   * Updates, creates, extends or clears session value based on the received authentication result.
+   * @param request Request instance.
+   * @param provider Provider that produced provided authentication result.
+   * @param authenticationResult Result of the authentication or login attempt.
+   * @param existingSessionValue Value of the existing session if any.
+   */
+  private async updateSessionValue(
+    request: KibanaRequest,
     {
       provider,
       authenticationResult,
-      existingSession,
-      isSystemRequest,
+      existingSessionValue,
     }: {
       provider: AuthenticationProvider;
       authenticationResult: AuthenticationResult;
-      existingSession: ProviderSession | null;
-      isSystemRequest: boolean;
+      existingSessionValue: Readonly<SessionValue> | null;
     }
   ) {
-    if (!existingSession && !authenticationResult.shouldUpdateState()) {
+    if (!existingSessionValue && !authenticationResult.shouldUpdateState()) {
+      return null;
+    }
+
+    // Provider can specifically ask to clear session by setting it to `null` even if authentication
+    // attempt didn't fail.
+    if (authenticationResult.shouldClearState()) {
```

Comment:
> Are there other scenarios I'm overlooking?

Yeah, I can think of the following scenario where we'd need this, e.g.:

* SAML/OIDC provider receives request with an existing session
* It detects that _both_ access and refresh token expired
* Now it needs to start a new IdP/OP handshake (return `AuthenticationResult.redirected()`) _and_ clear existing session

So in this case we didn't fail with `401` (since we can redirect user directly to the specific IdP instead, skipping login selector) and we didn't provide a new state (since we don't have any _and_ we want to regenerate SID anyway).

Do you think we can simplify that somehow?